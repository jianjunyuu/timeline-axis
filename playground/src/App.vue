<script setup lang="ts">
import { onMounted, ref, watchEffect } from 'vue'
import { TimelineAxis } from 'timeline-axis'

const timelineRef = ref<HTMLElement>()
// 初始化刻度
const initalValue = ref(2)
// 新增刻度值
const additionalValue = ref(0)

const axis = new TimelineAxis({
  width: window.innerWidth,
  height: 18,
})
onMounted(() => {
  timelineRef.value?.appendChild(axis.dom)

  watchEffect(() => {
    axis.setOptions({
      duration: Number(initalValue.value),
    })
  })
})

function handleScale(value: number) {
  if (value > 0)
    axis.zoomIn()
  else
    axis.zoomOut()
}

function addTo() {
  initalValue.value += Number(additionalValue.value)
  additionalValue.value = 0
}
</script>

<template>
  <div ref="timelineRef" class="timeline" />

  <input v-model="initalValue" aria-controls="">

  <div>
    <button @click="handleScale(-1)">
      click -
    </button>
    Scale {{ 0 }}
    <button @click="handleScale(1)">
      click +
    </button>
  </div>
  <input v-model="additionalValue">

  <button @click="addTo">
    add
  </button>
</template>

<style>
:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  font-weight: 400;

  color-scheme: light dark;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}
* {
  padding: 0;
  margin: 0;
}
</style>
