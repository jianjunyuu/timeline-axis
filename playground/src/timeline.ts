export function timeline() {

}
export type ScaleStepsValues = typeof ScaleSteps[number]
export const ScaleSteps = [1, 2, 3, 5, 10, 20, 30] as const
export function getScaleStep(t: number): ScaleStepsValues {
  let stepIndex = ScaleSteps.length - 1
  while (stepIndex >= 0) {
    if (t >= ScaleSteps[stepIndex])
      return ScaleSteps[stepIndex]

    stepIndex--
  }

  return ScaleSteps[0]
}

/**
 * min 1(40,25,15) 2(20) 3(23) 5(30,20) 10(30,20,15,12) max
 * 1s 2s 3s 5s 10s 20s 30s
 * 1m 2m 3m 5m 10m 20m 30m
 * 1h 2h 3h 6h 12h 24h
 */
export const StepWidthMap: Record<typeof ScaleSteps[number], number[]> = {
  1: [40, 25, 15],
  2: [20],
  3: [23],
  5: [30, 20],
  10: [30, 20, 15, 12],
  20: [20],
  30: [23],
}
