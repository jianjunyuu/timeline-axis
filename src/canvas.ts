import type { DrawOptions, Position, TickLabel, TickStyle } from './type'

export function draw(canvas: HTMLCanvasElement, stepWidth: number, stepSecond: number, options: DrawOptions) {
  clear(canvas)
  const tickCount = Math.floor(canvas.width / (stepWidth * 10))

  for (let index = 0; index <= tickCount; index++) {
    const x = index * stepWidth * 10
    if (index > 0) {
      drawLine(
        canvas,
        x,
        0,
        x,
        options.tick.height,
        options.tick.color,
        options.tick.width,
      )
      drawTimeLabel(canvas, index * stepSecond, {
        x: x + options.label.offset.x,
        y: options.label.offset.y,
      },
      options.label.style)
    }

    for (let k = 1; k < 10; k++) {
      const sX = x + k * stepWidth
      drawLine(
        canvas,
        sX,
        0,
        sX,
        options.smallTick.height,
        options.smallTick.color,
        options.smallTick.width,
      )
    }
  }
}
function clear(canvas: HTMLCanvasElement) {
  const context = canvas.getContext('2d')
  context?.clearRect(0, 0, canvas.width, canvas.height)
}
function drawLine(canvas: HTMLCanvasElement, beginX: number, beginY: number, endX: number, endY: number, color: Required<TickStyle>['color'], lineWidth: number) {
  const context = canvas.getContext('2d')
  if (context) {
    context.beginPath()
    context.moveTo(beginX, beginY)
    context.lineTo(endX, endY)
    context.strokeStyle = color
    context.lineWidth = lineWidth
    context.stroke()
  }
}

function fill(value: number) {
  return value < 10 ? `0${value}` : `${value}`
}
function drawTimeLabel(canvas: HTMLCanvasElement, time: number, position: Position, style: Required<TickLabel>['style']) {
  const date = new Date('2022-01-01 00:00:00')
  date.setTime(date.getTime() + time * 1000)

  const hour = date.getHours()
  const label = `${hour > 0 ? `${fill(hour)}:` : ''}${fill(date.getMinutes())}:${fill(date.getSeconds())}`
  const context = canvas.getContext('2d')
  if (context) {
    context.fillStyle = style
    context.fillText(label, position.x, position.y)
  }
}
