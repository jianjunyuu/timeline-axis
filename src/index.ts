export * from './timeline'

export * from './type'
