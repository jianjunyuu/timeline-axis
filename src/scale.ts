import { StepWidthMap, Steps } from './config'
import type { Scale } from './type'

export function getScaleStep(t: number): Scale['step'] {
  let stepIndex = Steps.length - 1
  while (stepIndex >= 0) {
    if (t >= Steps[stepIndex])
      return Steps[stepIndex]

    stepIndex--
  }

  return Steps[0]
}

export function getScales(duration: number) {
  const scales: Scale[] = []
  const [hour, minute, second] = getTime(duration)
  if (hour > 0)
    scales.push({ step: 1, width: 30, second: 60 * 60, text: `${1}时` })
  if (hour > 0 || minute > 0) {
    const value = getScaleStep(hour > 0 ? 59 : minute)
    for (let index = Steps.length - 1; index >= 0; index--) {
      const step = Steps[index]
      if (value >= step)
        scales.push(...StepWidthMap[step].map(width => ({ step, width, second: step * 60, text: `${step}分` })).reverse())
    }
  }

  const value = getScaleStep(minute > 0 ? 59 : second)
  for (let index = Steps.length - 1; index >= 0; index--) {
    const step = Steps[index]
    if (value >= step)
      scales.push(...StepWidthMap[step].map(width => ({ step, width, second: step, text: `${step}秒` })).reverse())
  }

  return scales
}
function getTime(duration: number): [number, number, number] {
  const date = new Date('2022-01-01 00:00:00')
  date.setTime(date.getTime() + duration * 1000)
  return [date.getHours(), date.getMinutes(), date.getSeconds()]
}
