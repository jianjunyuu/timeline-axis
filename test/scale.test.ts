import { describe, expect, it } from 'vitest'
import { getScaleStep, getScales } from '../src/scale'

describe('get step', () => {
  it.each([
    [0, 1],
    [1, 1],
    [2, 2],
    [3, 3],
    [4, 3],
    [5, 5],
    [6, 5],
    [10, 10],
    [11, 10],
    [20, 20],
    [21, 20],
    [30, 30],
    [99, 30],
  ])('getScaleStep(%i) -> %i', (value, expected) => {
    expect(getScaleStep(value)).toBe(expected)
  })
})

describe('get scales by duration(second)', () => {
  it.each([
    [1],
    [2],
    [3],
    [4],
    [5],
    [6],
    [9],
    [10],
    [19],
    [20],
    [30],
    [59],
    [60],
    [61],
    [90],
    [91],
    [120],
    [121],
    [30 * 60 + 10],
    [60 * 60],
    [60 * 60 + 1],
  ])('getScales(%i) -> %j', (value) => {
    expect(getScales(value)).toMatchSnapshot()
  })
})
