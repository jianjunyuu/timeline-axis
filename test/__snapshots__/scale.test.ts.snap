// Vitest Snapshot v1

exports[`get scales by duration(second) > getScales(1) -> %j 1`] = `
[
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(2) -> %j 1`] = `
[
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(3) -> %j 1`] = `
[
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(4) -> %j 1`] = `
[
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(5) -> %j 1`] = `
[
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(6) -> %j 1`] = `
[
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(9) -> %j 1`] = `
[
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(10) -> %j 1`] = `
[
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(19) -> %j 1`] = `
[
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(20) -> %j 1`] = `
[
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(30) -> %j 1`] = `
[
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(59) -> %j 1`] = `
[
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(60) -> %j 1`] = `
[
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(61) -> %j 1`] = `
[
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(90) -> %j 1`] = `
[
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(91) -> %j 1`] = `
[
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(120) -> %j 1`] = `
[
  {
    "second": 120,
    "step": 2,
    "text": "2分",
    "width": 20,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(121) -> %j 1`] = `
[
  {
    "second": 120,
    "step": 2,
    "text": "2分",
    "width": 20,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(1810) -> %j 1`] = `
[
  {
    "second": 1800,
    "step": 30,
    "text": "30分",
    "width": 23,
  },
  {
    "second": 1200,
    "step": 20,
    "text": "20分",
    "width": 20,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 12,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 15,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 20,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 30,
  },
  {
    "second": 300,
    "step": 5,
    "text": "5分",
    "width": 20,
  },
  {
    "second": 300,
    "step": 5,
    "text": "5分",
    "width": 30,
  },
  {
    "second": 180,
    "step": 3,
    "text": "3分",
    "width": 23,
  },
  {
    "second": 120,
    "step": 2,
    "text": "2分",
    "width": 20,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 30,
    "step": 30,
    "text": "30秒",
    "width": 23,
  },
  {
    "second": 20,
    "step": 20,
    "text": "20秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 12,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 15,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 20,
  },
  {
    "second": 10,
    "step": 10,
    "text": "10秒",
    "width": 30,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 20,
  },
  {
    "second": 5,
    "step": 5,
    "text": "5秒",
    "width": 30,
  },
  {
    "second": 3,
    "step": 3,
    "text": "3秒",
    "width": 23,
  },
  {
    "second": 2,
    "step": 2,
    "text": "2秒",
    "width": 20,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(3600) -> %j 1`] = `
[
  {
    "second": 3600,
    "step": 1,
    "text": "1时",
    "width": 30,
  },
  {
    "second": 1800,
    "step": 30,
    "text": "30分",
    "width": 23,
  },
  {
    "second": 1200,
    "step": 20,
    "text": "20分",
    "width": 20,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 12,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 15,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 20,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 30,
  },
  {
    "second": 300,
    "step": 5,
    "text": "5分",
    "width": 20,
  },
  {
    "second": 300,
    "step": 5,
    "text": "5分",
    "width": 30,
  },
  {
    "second": 180,
    "step": 3,
    "text": "3分",
    "width": 23,
  },
  {
    "second": 120,
    "step": 2,
    "text": "2分",
    "width": 20,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;

exports[`get scales by duration(second) > getScales(3601) -> %j 1`] = `
[
  {
    "second": 3600,
    "step": 1,
    "text": "1时",
    "width": 30,
  },
  {
    "second": 1800,
    "step": 30,
    "text": "30分",
    "width": 23,
  },
  {
    "second": 1200,
    "step": 20,
    "text": "20分",
    "width": 20,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 12,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 15,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 20,
  },
  {
    "second": 600,
    "step": 10,
    "text": "10分",
    "width": 30,
  },
  {
    "second": 300,
    "step": 5,
    "text": "5分",
    "width": 20,
  },
  {
    "second": 300,
    "step": 5,
    "text": "5分",
    "width": 30,
  },
  {
    "second": 180,
    "step": 3,
    "text": "3分",
    "width": 23,
  },
  {
    "second": 120,
    "step": 2,
    "text": "2分",
    "width": 20,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 15,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 25,
  },
  {
    "second": 60,
    "step": 1,
    "text": "1分",
    "width": 40,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 15,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 25,
  },
  {
    "second": 1,
    "step": 1,
    "text": "1秒",
    "width": 40,
  },
]
`;
